### PR TITLE
feat: toml config

### PR DESCRIPTION
**Pull request by r3tr0ananas**

### CODE REVIEW

**Comment by THEGOLDENPRO** [comment]
Nice, going to do some changes. (just in-case you have outstanding commits)

**Comment by THEGOLDENPRO** [comment]
dammit, with serde there's no way to generate toml with comments?

**Comment by THEGOLDENPRO** [comment]
I think I might just create a template then because I also want it in a specific format.
image

**Comment by r3tr0ananas** [comment]
image
It looks like this for me whar

**Comment by r3tr0ananas** [comment]
Also i forgot to do `image.loading`

**Comment by THEGOLDENPRO** [comment]
> image It looks like this for me whar

I changed it, I wanted keys like `info_box_toggle` to be separate like `info_box.toggle`.

**Comment by r3tr0ananas** [comment]
Ohhh, i didn't see the dots, i thought they were underscores.

**Comment by THEGOLDENPRO** [comment]
ohhhh  question mark error handling ohhh  rust.

image

**Comment by THEGOLDENPRO** [comment]
truely unbreakable...

image

File: src/config.rs [hunk 1]
```diff
@@ -66,8 +66,8 @@ impl Config {
         let config_dir_path = local_dir.join("cloudy").join("roseate");
 
         if !config_dir_path.exists() {
-            fs::create_dir_all(config_dir_path.clone())?;
-            if let Err(err) = fs::create_dir_all(config_dir_path.clone()) {
+            fs::create_dir_all(&config_dir_path)?;
+            if let Err(err) = fs::create_dir_all(&config_dir_path) {
```
Comment (r3tr0ananas) [inline src/config.rs:70]: image

Comment (THEGOLDENPRO) [reply src/config.rs]: Don't make me remove Arc<T> from the `image_bytes` field in the `Image` struct. (everywhere `image.clone()` is used the entire image loaded in memory will duplicate and WE'VE USED IMAGE.CLONE LIKE EVERYWHERE! 💀)

**Comment by THEGOLDENPRO** [comment]
Alright I'm done with my commits.

It's just that right now `.expect()` you chained after `Key::from_name` will cause roseate to panic and crash if the user set a keybind that is invalid for egui.

Instead of crashing I would like it to send a toast to the user stating that the key bind is invalid but before we can do that I think we first need to create a separate struct for managing toasts so we just pass a reference of that to functions instead of passing `&mut toasts` everywhere. Also I planned to move the `log_and_toasts` function to that struct.

**Comment by THEGOLDENPRO** [comment]
perfect

image

**Comment by THEGOLDENPRO** [comment]
One problem with this that I'm realizing now: If we release a new version of roseate that has some new toml fields, `toml::from_str::<Config>(&value)?` will fail to deserialize it cause all the toml keys to fallback to [default values](https://github.com/cloudy-org/roseate/blob/feat/toml-config/src/main.rs#L117).

This isn't good, is there a way to get `toml::from_str` to not do that and just return the default value for that missing toml field?

**Comment by THEGOLDENPRO** [comment]
Maybe `#[serde(default)]`  but fuck this I'm going to sleep.

**Comment by THEGOLDENPRO** [comment]
Idk try it, remove some toml key in your config.toml and see if you get an error. If you do then it's not working.

**Comment by r3tr0ananas** [comment]
I tested it and it works.

**Comment by THEGOLDENPRO** [comment]
This doesn't seem to be working correctly, like yes it is defaulting but it's not defaulting to our values. For example all keybinds, since they are a `String`, default to an empty strings (""). The arrow is what we are getting from `&config.keybinds.info_box.toggle`.

Someone else is also experiencing this [issue](https://users.rust-lang.org/t/serde-default-versus-impl-default/66773).

image

**Comment by r3tr0ananas** [comment]
This works fine for me.

image

**Comment by THEGOLDENPRO** [comment]
Idk I tested it by adding a mistyped version of info box toggle:

```toml
info_box.togle = "I"
```

Maybe try that

**Comment by THEGOLDENPRO** [comment]
Yeah this won't work, the serde docs even say it only works on structs, so I'm guessing it works when the entire struct is missing but not a value in the struct which sucks.

I think we should just use a get method at this point but if we still want nicely structured configs in the codebase figment looks a little promising.

https://www.reddit.com/r/rust/s/0SUBf7FRFA

**Comment by r3tr0ananas** [comment]
I know how i could fix it.

I'll push that fix in an hour or two.

**Comment by r3tr0ananas** [comment]
So this should be fixed what you experienced.

**Comment by THEGOLDENPRO** [comment]
Yeah I knew this was a way to fix it but *ughh it's ugly* but ughh it works. They are planning to add support for something like `default = false` and etc but they still haven't added it.

I think I'm just gonna move these structs and functions into separate files under some `config/mod.rs` directory or something.

**Comment by r3tr0ananas** [comment]
I put all structs into a folder named models.

**Comment by r3tr0ananas** [comment]
@THEGOLDENPRO at this point you can make a new devlog.

**Comment by THEGOLDENPRO** [comment]
when I'm done with my multi-threaded journey

**Comment by THEGOLDENPRO** [comment]
I'm not sure about marginal allowance yet 🤔

I was thinking of humanizing that config more by hiding it behind something like this:
```toml
[image.optimizations]
downsampling.mode = "strict"
```

but I'm still thinking about the values these strings will correspond to... well I'm still thinking about EVERYTHING so I'm not sure yet.

**Comment by r3tr0ananas** [comment]
tbh i would just keep it like that and add a note: if you don't know what you're doing then don't touch it.

File: assets/config.template.toml [hunk 2]
```diff
@@ -17,6 +17,16 @@ gui.lazy_loading = true
 # recommended to set this to "false" as it will lead to the GUI freezing each time you try to zoom in/out.
 dynamic.lazy_loading = true
 
+[ui_settings]
```
Comment (THEGOLDENPRO) [inline assets/config.template.toml:20]: we can shorten this down to just `ui`

Comment (THEGOLDENPRO) [reply assets/config.template.toml:20]: already done in my commit btw, I did a lot of refactoring so be ready for merge conflicts if you changed anything :)

**Comment by THEGOLDENPRO** [comment]
> tbh i would just keep it like that and add a note: if you don't know what you're doing then don't touch it. 

I'm gonna leave marginal allowance out for now as the image optimization stuff is still very work in progress and I have a lot of changes for this portion of the code in some uncommitted commits so this config won't last long anyways.

Also I'll be merging this now.

**Comment by r3tr0ananas** [comment]
@THEGOLDENPRO when new dev log dropping

**Pull request merged**